cmd/collectors/ems/ems.go

```diff
@@ -574,128 +622,228 @@ func (e *Ems) HandleResults(result gjson.Result, prop map[string][]*emsProp) (ma
-								Msg("Unable to set float key on metric")
+
+					for _, metric := range p.Metrics {
+						metr, ok := mx.GetMetrics()[metric.Name]
+						if !ok {
+							if metr, err = mx.NewMetricFloat64(metric.Name); err != nil {
+								e.Logger.Error().Err(err).
+									Str("name", metric.Name).
+									Msg("NewMetricFloat64")
+								continue
+							}
+							metr.SetExportable(metric.Exportable)
 						}
-					} else {
-						// this code will not execute as ems only support events metric
-						f := instanceData.Get(metric.Name)
-						if f.Exists() {
-							if err = metr.SetValueFloat64(instance, f.Float()); err != nil {
+						if metric.Name == "events" {
+							if err = metr.SetValueFloat64(instance, 1); err != nil {
 								e.Logger.Error().Err(err).Str("key", metric.Name).Str("metric", metric.Label).
 									Msg("Unable to set float key on metric")
 							}
+						} else if metric.Name == "timestamp" {
+							if err = metr.SetValueFloat64(instance, float64(time.Now().UnixMicro())); err != nil {
+								e.Logger.Error().Err(err).Str("key", metric.Name).Str("metric", metric.Label).
+									Msg("Unable to set timestamp on metric")
+							}
+						} else {
+							// this code will not execute as ems only support events metric
+							f := instanceData.Get(metric.Name)
+							if f.Exists() {
+								if err = metr.SetValueFloat64(instance, f.Float()); err != nil {
+									e.Logger.Error().Err(err).Str("key", metric.Name).Str("metric", metric.Label).
+										Msg("Unable to set float key on metric")
+								}
+							}
 						}
 					}
 				}
 			}
+			if !isMatch {
+				mx.RemoveInstance(instanceKey)
+				return true
+			}
+			count += instanceLabelCount
 		}
-		if !isMatch {
-			mx.RemoveInstance(instanceKey)
-			return true
-		}
-		count += instanceLabelCount
 		return true
 	})
 	return m, count
 }
 
+func (e *Ems) getInstanceKeys(p *emsProp, instanceData gjson.Result) string {
+	var instanceKey string
+	// extract instance key(s)
+	for _, k := range p.InstanceKeys {
+		value := parseProperties(instanceData, k)
+		if value.Exists() {
+			instanceKey += value.String()
+		} else {
+			e.Logger.Warn().Str("key", k).Msg("skip instance, missing key")
+			break
+		}
+	}
+	return instanceKey
+}
+
+func (e *Ems) updateMatrix() {
+	var (
+		ok   bool
+		val  float64
+		metr matrix.Metric
+	)
+
+	tempMap := make(map[string]*matrix.Matrix)
+	// store the bookend ems metric in tempMap
+	for _, issuingEmsList := range e.bookendEmsMap {
+		for _, issuingEms := range issuingEmsList {
+			if mx, exist := e.Matrix[issuingEms]; exist {
+				tempMap[issuingEms] = mx
+			}
+		}
+	}
+
+	// remove all ems matrix except parent object
+	mat := e.Matrix[e.Object]
+	e.Matrix = make(map[string]*matrix.Matrix)
+	e.Matrix[e.Object] = mat
+
+	for issuingEms, mx := range tempMap {
+		if metr, ok = mx.GetMetrics()["events"]; !ok {
+			e.Logger.Error().
+				Str("name", "events").
+				Msg("NewMetricFloat64")
```

Comment:
log msg doesn't match code. Does `Str("name", "events")` help here? The logger will already tell us this is from the ems.go code. maybe `Msg("failed to get events metric")` or something like that is clearer

Reply:
done